File: problems/allpairspath/allpairspath.cpp
#include <algorithm>
#include <iostream>
#include <limits>

using namespace std;

#define INF std::numeric_limits<double>::infinity()
#define NINF -INF

double inf_add(double a, double b) {
	// a or b being INF means there's no edge, so nothing to add
	if (a == INF || b == INF) {
		return INF;
	}
	return a + b;
}

int main()
{
	int n, m, q;
	int i, j, k;
	int u, v, w;
	double dist;

	// graph of maximum size
	double graph[150][150];

	cin >> n >> m >> q;
	while (n != 0) {

		// Reset distances
		for (i = 0; i < n; i++) {
			for (j = 0; j < n; j++) {
				graph[i][j] = (i == j) ? 0.0 : INF;
			}
		}

		// Load edges
		for (i = 0; i < m; i++) {
			cin >> u >> v >> w;
			graph[u][v] = min(graph[u][v], (double) w);
		}

		// Shortest paths through k
		for (k = 0; k < n; k++) {
			for (i = 0; i < n; i++) {
				for (j = 0; j < n; j++) {
					dist = min(inf_add(graph[i][k], graph[k][j]), graph[i][j]);
					if (i == j && dist < 0.0) {
						dist = NINF;
					}
					graph[i][j] = dist;
				}
			}
		}

		// Queries
		for (i = 0; i < q; i++) {
			cin >> u >> v;
			// cout << u << ' ' << v << '\n';
			if (graph[u][v] == INF) {
				cout << "Impossible" << std::endl;
			}
			else if (graph[u][v] == -INF) {
				cout << "-Infinity" << std::endl;
			}
			else {
				cout << ((int) graph[u][v]) << std::endl;
			}
		}
		cout << std::endl;

		cin >> n >> m >> q;
	}
	return 0;
}

File: problems/allpairspath/allpairspath.py
safesum = lambda *v: None if None in v else sum(v)

def safemin(*v):
	try:
		return min(filter(lambda x: x is not None, v))
	except:
		return None

neginf = float('-inf')

# n=#nodes, m=#edges, q=#queries
n, m, q = map(int, input().split())
while n != 0:

	# Adjacency matrix
	dist = [[None]*n for _ in range(n)]
	for i in range(n):
		dist[i][i] = 0
	for _ in range(m):
		u, v, w = map(int, input().split())
		if w < 0:
			dist[u][v] = -1
			continue

		if u == v:
			dist[u][v] = min(dist[u][v], w)
		else:
			dist[u][v] = w

	for k in range(n):
		for i in range(n):
			for j in range(n):
				if i == j and dist[i][j] < 0:
					dist[i][j] = neginf
				else:
					dist[i][j] = safemin(safesum(dist[i][k], dist[k][j]), dist[i][j])

	for _ in range(q):
		a, b = map(int, input().split())
		if dist[a][b] == neginf:
			print("-Infinity")
		elif dist[a][b] == None:
			print("Impossible")
		else:
			print(dist[a][b])
	print()

	n, m, q = map(int, input().split())
